sdk/helper/keysutil/policy.go

```diff
@@ -2392,3 +2395,191 @@ func wrapTargetPKCS8ForImport(wrappingKey *rsa.PublicKey, preppedTargetKey []byt
 	wrappedKeys := append(ephKeyWrapped, targetKeyWrapped...)
 	return base64.StdEncoding.EncodeToString(wrappedKeys), nil
 }
+
+func (p *Policy) CreateCSR(keyVersion int, csrTemplate *x509.CertificateRequest) ([]byte, error) {
+	if !p.Type.SigningSupported() {
+		return nil, errutil.UserError{Err: fmt.Sprintf("key type '%s' does not support signing", p.Type)}
+	}
+
+	keyEntry, err := p.safeGetKeyEntry(keyVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	if keyEntry.IsPrivateKeyMissing() {
+		return nil, errutil.UserError{Err: "private key not imported for key version selected"}
+	}
+
+	csrTemplate.Signature = nil
+	csrTemplate.SignatureAlgorithm = x509.UnknownSignatureAlgorithm
+
+	key, err := p.getPrivateKey(&keyEntry)
+	if err != nil {
+		return nil, err
+	}
+
+	csrBytes, err := x509.CreateCertificateRequest(rand.Reader, csrTemplate, key)
+	if err != nil {
+		return nil, fmt.Errorf("could not create the CSR: %w", err)
+	}
+
+	pemCSR := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE REQUEST",
+		Bytes: csrBytes,
+	})
+
+	return pemCSR, nil
+}
+
+func (p *Policy) getPrivateKey(keyEntry *KeyEntry) (crypto.Signer, error) {
+	switch p.Type {
+	case KeyType_ECDSA_P256, KeyType_ECDSA_P384, KeyType_ECDSA_P521:
+		curve, err := p.getECDSAKeyCurve()
+		if err != nil {
+			return nil, err
+		}
+		key := &ecdsa.PrivateKey{
+			PublicKey: ecdsa.PublicKey{
+				Curve: curve,
+				X:     keyEntry.EC_X,
+				Y:     keyEntry.EC_Y,
+			},
+			D: keyEntry.EC_D,
+		}
+		return key, nil
+	case KeyType_ED25519:
+		if p.Derived {
+			return nil, errutil.UserError{Err: "operation not supported on keys with derivation enabled"}
+		}
+		return ed25519.PrivateKey(keyEntry.Key), nil
+	case KeyType_RSA2048, KeyType_RSA3072, KeyType_RSA4096:
+		return keyEntry.RSAKey, nil
+	default:
+		return nil, errutil.InternalError{Err: fmt.Sprintf("selected key type '%s' does not support signing", p.Type.String())}
+	}
+}
+
+func (p *Policy) getECDSAKeyCurve() (elliptic.Curve, error) {
+	switch p.Type {
+	case KeyType_ECDSA_P256:
+		return elliptic.P256(), nil
+	case KeyType_ECDSA_P384:
+		return elliptic.P384(), nil
+	case KeyType_ECDSA_P521:
+		return elliptic.P521(), nil
+	default:
+		return nil, errutil.InternalError{Err: fmt.Sprintf("key type '%s' does not have a curve parameter", p.Type)}
+	}
+}
+
+func (p *Policy) ValidateAndPersistCertificateChain(ctx context.Context, storage logical.Storage, keyVersion int, certificateChain []*x509.Certificate) error {
+	if len(certificateChain) == 0 {
+		return errutil.UserError{Err: "expected at least a certificate in the certificate chain"}
+	}
+
+	if certificateChain[0].BasicConstraintsValid && certificateChain[0].IsCA {
```

Comment:
Hmm...

Part of me wants to walk this back: what if a key in Transit is for a CA certificate, not a leaf certificate?

In that case, I think the check would be:

```
if len(certChain) > 1 && certChain[0].BasicConstraintsValid && certChain[0].IsCA && (!certChain[-1].BCV || !certChain[-1].IsCA) {
  ...
}
```

i.e., our first item is a CA and our last item is not, implying the cert chain was presented in incorrect order... 

Thoughts?

Reply:
> Part of me wants to walk this back: what if a key in Transit is for a CA certificate, not a leaf certificate?

If this is the case, the call below would have to be somewhat different and check if there's a match between the key version in transit and a key in the chain. 
> 	valid, err := p.validateCertificateKeyMatch(keyVersion, certificateChain[0].PublicKeyAlgorithm, certificateChain[0].PublicKey)

With that, I'm not sure if I'm understanding your comment. Is it suggesting an additional check to make sure there's a CA certificate in the last position?

Reply:
I was envisioning you'd always have the leaf-most CA's chain here, never a child CA/leaf cert if you had the CA's key. i.e., root -> ... -> leaf CA. ..

But thinking about this more, I'm not sure you'd have much value in a CA certificate that doesn't issue leaves (and thus, would be better suited for the PKI engine)... Unless you were issuing a type of certificate that Transit supported but PKI didn't and/or implementing CA software yourself. But then you'd probably have some other CA cert storage location. The one exception to this is a lot of self-signed service certs default to `IsCA=True` even if they're just TLS cets, so I could see that happening, but they should be able to change their self-signed creation step to set `IsCA=false` and then import it. 

So, let's keep this to leaves only is my thought!